src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java

```diff
@@ -493,7 +495,9 @@ private void stopConnection(String errorMessage) {
             redirectAccessTokenProvider = null;
             connectionId = null;
             transportEnum = TransportEnum.ALL;
-            this.headers.remove("Authorization");
+            if (tokenSetByNegotiate) {
+                this.headers.remove("Authorization");
```

Comment:
Why are we removing the header in the first place? I don't think we do anything like this for the other clients.
I looked at when we added it and the issue was that we weren't saving the original `accessTokenProvider`.

I think all of our clients (at least C# and Java) have a problem if someone sets the `Authorization` header via options and then either provides an `accessTokenProvider` or uses the service. Because the original `Authorization` header will be replaced and never used again. Which could cause the client to never be able to reconnect if the server is doing custom auth based on the original `Authorization` header. We should probably consider a broader fix for that class of issues.

Reply:
We should not be adding the result of `redirectAccessTokenProvider` directly to `this.headers` only in negotiate and relying on the headers dictionary to keep the state for the duration of the connection. Instead, we should have a prepare requests method that calls `redirectAccessTokenProvider` for every request and re-adds the token to the to a dictionary used only for that request.

Reply:
We actually had the correct behavior in the LongPolling transport. Before any request we we call an `updateHeaderToken` method. For the Websockets transport we only care about updating the token right before we connect, so we shouldn't need extra logic in the transport, right? Updating the token in the `HubConnection` 's start should be enough.

Reply:
The main issue still is how do we want to handle users explicitly setting the Auth header, because currently they wouldn't be able to use the SignalR service. The reported issue would be fixed because we don't clear the user set auth header when restarting a connection

Reply:
It's just difficult to manually undo changes you make to a dictionary. It's better to shallow-copy the original dictionary, add the entry you want, and then make the single websocket/longpolling/whatever request with the shallow-copied dictionary. That way you're sure you never end up in a state were the original dictionary is in an invalid state.